packages/dds/map/src/test/mocha/summary.spec.ts

```diff
@@ -0,0 +1,186 @@
+	assert(second !== undefined, "second root dir should be present");
+
+	// Check number of keys.
+	assert.strictEqual(
+		first.size,
+		second.size,
+		`Number of keys not same: Number of keys ` +
+			`in first at path ${first.absolutePath}: ${first.size} and in second at path ${second.absolutePath}: ${second.size}`,
+	);
+
+	// Check key/value pairs in both directories.
+	for (const key of first.keys()) {
+		assert.strictEqual(
+			first.get(key),
+			second.get(key),
+			`Key not found or value not matching ` +
+				`key: ${key}, value in dir first at path ${first.absolutePath}: ${first.get(
+					key,
+				)} and in second at path ${second.absolutePath}: ${second.get(key)}`,
+		);
+	}
+
+	// Check for number of subdirectores with both directories.
+	assert(first.countSubDirectory !== undefined && second.countSubDirectory !== undefined);
+	assert.strictEqual(
+		first.countSubDirectory(),
+		second.countSubDirectory(),
+		`Number of subDirectories not same: Number of subdirectory in ` +
+			`first at path ${first.absolutePath}: ${first.countSubDirectory()} and in second` +
+			`at path ${second.absolutePath}: ${second.countSubDirectory()}`,
+	);
+
+	// Check for consistency of subdirectores with both directories.
+	for (const [name, subDirectory1] of first.subdirectories()) {
+		const subDirectory2 = second.getSubDirectory(name);
+		assert(
+			subDirectory2 !== undefined,
+			`SubDirectory with name ${name} not present in second directory`,
+		);
+		assertEventualConsistencyCore(subDirectory1, subDirectory2);
+	}
+
+	// Check for consistency of subdirectories ordering of both directories
+	const firstSubdirNames = [...first.subdirectories()].map(([dirName, _]) => dirName);
+	const secondSubdirNames = [...second.subdirectories()].map(([dirName, _]) => dirName);
+	assert.deepStrictEqual(firstSubdirNames, secondSubdirNames);
+}
+
+function takeSnapshot(directory: SharedDirectory, fileName: string): string {
+	const data = serialize(directory);
+	if (!existsSync(fileName)) {
+		writeFileSync(fileName, data);
+	}
+	const pastData = readFileSync(fileName, "utf8");
+	assert.equal(data, pastData, `snapshots are different on test "${fileName}"`);
+	return data;
+}
+
+async function loadSharedDirectory(
+	id: string,
+	serializedSnapshot: string,
+): Promise<SharedDirectory> {
+	const containerRuntimeFactory = new MockContainerRuntimeFactory();
+	const dataStoreRuntime = new MockFluidDataStoreRuntime();
+	const containerRuntime = containerRuntimeFactory.createContainerRuntime(dataStoreRuntime);
+	const services = {
+		deltaConnection: dataStoreRuntime.createDeltaConnection(),
+		// eslint-disable-next-line @typescript-eslint/no-unsafe-argument
+		objectStorage: new MockStorage(JSON.parse(serializedSnapshot)),
+	};
+	const sharedDirectory = new SharedDirectory(id, dataStoreRuntime, DirectoryFactory.Attributes);
+	await sharedDirectory.load(services);
+	return sharedDirectory;
+}
+
+function generateTestScenarios(): TestScenario[] {
```

Comment:
other than two comments above, this generally looks like a good start.

I think before we close the work item, it'd be good to invest a bit on infrastructure for *back-compat* of snapshots. E.g. things like "Can we still load and collaborate with some specific document that was generated using a previous format, even if we no longer write that format?"

Right now we have some minor coverage of that by inlining snapshot contents into map/directory tests, but having a more systematic approach will pay off long-term. Could you look into/rationalize how we might want to organize these tests? I suspect we're probably missing coverage for some features of map/directory snapshots, given we've made bugfixes which impact the format in small ways a few times in the last year.

Reply:
Added a snapshot written manually in the old format, only testing the read-compatibility